x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go

```diff
@@ -0,0 +1,223 @@
+
+	for _, db := range dbs {
+		moduleFields := common.MapStr{
+			"database": common.MapStr{
+				"id":   db.id,
+				"name": db.name,
+			},
+		}
+		metricsetFields := common.MapStr{}
+
+		spaceUsage, err := m.getLogSpaceUsageForDb(db.name)
+		if err != nil {
+			reporter.Error(err)
+		} else {
+			metricsetFields["space_usage"] = spaceUsage
+		}
+
+		stats, err := m.getLogStats(db)
+		if err != nil {
+			reporter.Error(err)
+		} else {
+			metricsetFields["stats"] = stats
+		}
+
+		// Both log space and log size are available, so report both
+		if isReported := reporter.Event(mb.Event{
+			ModuleFields:    moduleFields,
+			MetricSetFields: metricsetFields,
+		}); !isReported {
+			m.log.Debug("event not reported")
+		}
+
+	}
+}
+
+// Close the connection to the server at the engine level
+func (m *MetricSet) Close() error {
+	return m.db.Close()
+}
+
+func (m *MetricSet) getLogSpaceUsageForDb(dbName string) (result common.MapStr, err error) {
+
+	rows, err := m.db.Query(fmt.Sprintf(`USE %s; SELECT * FROM sys.dm_db_log_space_usage;`, dbName))
+	if err != nil {
+		return
+	}
+	defer closeRows(rows)
+
+	// According to docs "All transaction log files are combined." Means,
+	// a single result is returned
+	for rows.Next() {
+		var res logSpace
+		if err = rows.Scan(&res.id, &res.totalLogSizeInBytes, &res.usedLogSpaceInBytes, &res.usedLogSpaceInPercent,
+			&res.logSpaceInBytesSinceLastBackup); err != nil {
+			return
+		}
+
+		result = common.MapStr{
+			"total": common.MapStr{
+				"bytes": res.totalLogSizeInBytes,
+			},
+			"used": common.MapStr{
+				"bytes": res.usedLogSpaceInBytes,
+				"pct":   res.usedLogSpaceInPercent,
+			},
+			"since_last_backup": common.MapStr{
+				"bytes": res.logSpaceInBytesSinceLastBackup,
+			},
+		}
+	}
+
+	return
+}
+func (m *MetricSet) getLogStats(db dbInfo) (result common.MapStr, err error) {
+
+	rows, err := m.db.Query(fmt.Sprintf(`USE %s; SELECT database_id,total_log_size_mb,active_log_size_mb,log_backup_time,log_since_last_log_backup_mb,log_since_last_checkpoint_mb,log_recovery_size_mb FROM sys.dm_db_log_stats(%d);`, db.name, db.id))
+	if err != nil {
+		return
+	}
+	defer closeRows(rows)
+
+	for rows.Next() {
+		var res logStats
+		if err = rows.Scan(&res.databaseID, &res.sizeMB, &res.activeSizeMB, &res.backupTime, &res.sinceLastBackupMB, &res.sinceLastCheckpointMB, &res.recoverySizeMB); err != nil {
+			return
+		}
+
+		result = common.MapStr{
+			"total_size": common.MapStr{
+				"bytes": res.sizeMB * 1048576,
+			},
+			"active_size": common.MapStr{
+				"bytes": res.activeSizeMB * 1048576,
+			},
+			"backup_time": res.backupTime,
+			"since_last_checkpoint": common.MapStr{
+				"bytes": res.sinceLastCheckpointMB * 1048576,
+			},
+			"recovery_size": common.MapStr{
+				"bytes": res.recoverySizeMB,
+			},
+		}
+
+		if res.sinceLastBackupMB != nil {
+			result.Put("since_last_backup.bytes", *res.sinceLastBackupMB*1048576)
+		}
+		// According to docs "All transaction log files are combined." Means,
```

Comment:
This comment should be on line 165, perhaps not need if only 1 row is fetched. See other comment for content.